### PR TITLE
Add target mode to sfputil firmware

### PR DESCRIPTION
**Pull request by AnoopKamath**
#### What I did
Add 'target' option to sfputil CLI. This target mode will be set in the vendor custom registers in the module page 0h Byte 60, which will direct the right path to upgrade remote module

#### How I did it
Add target option in the sfputil file and add related vendor specific apis in the CMIS code.

#### How to verify it
Use remote upgraded module and execute the sfputil firmware command.

#### Previous command output (if the output of a command-line utility has changed)

```
root@sonic:/home/cisco# sfputil firmware ?
Usage: sfputil firmware [OPTIONS] COMMAND [ARGS]...

  Download/Upgrade firmware on the transceiver

Options:
  --help  Show this message and exit.

Commands:
  commit    Commit the running firmware
  download  Download firmware on the transceiver
  run       Run the firmware with default mode=0
  unlock    Unlock the firmware download feature via CDB host password
  upgrade   Upgrade firmware on the transceiver
```
#### New command output (if the output of a command-line utility has changed)

```
root@sonic:/home/cisco# sfputil firmware ?
Usage: sfputil firmware [OPTIONS] COMMAND [ARGS]...

  Download/Upgrade firmware on the transceiver

Options:
  --help  Show this message and exit.

Commands:
  commit    Commit the running firmware
  download  Download firmware on the transceiver
  run       Run the firmware with default mode=0
  target    Select target end for firmware download 0-(local) 1-(remote-A)...
  unlock    Unlock the firmware download feature via CDB host password
  upgrade   Upgrade firmware on the transceiver
root@sonic:/home/cisco#

root@sonic:/home/cisco# sfputil firmware target --help
Usage: sfputil firmware target [OPTIONS] PORT_NAME TARGET

  Select target end for firmware download 0-(loca)

  1-(remote-A)

  2-(remote-B)

Options:
  --help  Show this message and exit.
root@sonic:/home/cisco#

```

### CODE REVIEW

File: sfputil/main.py [hunk 1]
```diff
@@ -1526,6 +1526,48 @@ def version():
...
+# 'target' subcommand
+@firmware.command()
+@click.argument('port_name', required=True, default=None)
+@click.argument('target', type=click.INT, required=True, default=None)
+def target(port_name, target):
+    """Select target end for firmware download 0-(local) \n
+                                               1-(remote-A) \n
+                                               2-(remote-B)
+    """
+    if target > 2:
```
Comment (prgeor) [inline sfputil/main.py:1538]: @AnoopKamath can you use click specific argument checking?

File: sfputil/main.py [hunk 2]
```diff
@@ -1526,6 +1526,48 @@ def version():
...
+    try:
+        api = sfp.get_xcvr_api()
+    except NotImplementedError:
+        click.echo("This functionality is currently not implemented for this module")
+        sys.exit(ERROR_NOT_IMPLEMENTED)
+
+    try:
+        status = api.set_firmware_download_target_end(target)
+    except AttributeError:
+        click.echo("This functionality is currently not implemented for this module")
```
Comment (prgeor) [inline sfputil/main.py:1563]: @AnoopKamath could you print the port name

File: sfputil/main.py [hunk 3]
```diff
@@ -1526,6 +1526,48 @@ def version():
...
+        sys.exit(EXIT_FAIL)
+
+    if not is_sfp_present(port_name):
+       click.echo("{}: SFP EEPROM not detected\n".format(port_name))
+       sys.exit(EXIT_FAIL)
+
+    try:
+        api = sfp.get_xcvr_api()
+    except NotImplementedError:
+        click.echo("This functionality is currently not implemented for this module")
```
Comment (prgeor) [inline sfputil/main.py:1557]: @AnoopKamath could you print port name?

**Comment by prgeor** [comment]
@AnoopKamath please add unit test to have code coverage pass

File: sfputil/main.py (hunk 1 above)
Comment (AnoopKamath) [reply sfputil/main.py]: Taken care.

```
root@sonic:/home/cisco# sfputil firmware target Ethernet0 3
Usage: sfputil firmware target [OPTIONS] PORT_NAME TARGET
Try "sfputil firmware target --help" for help.

Error: Invalid value for "TARGET": 3 is not in the valid range of 0 to 2.
```

File: sfputil/main.py (hunk 3 above)
Comment (AnoopKamath) [reply sfputil/main.py]: taken care

File: sfputil/main.py (hunk 2 above)
Comment (AnoopKamath) [reply sfputil/main.py]: taken care

**Comment by AnoopKamath** [comment]
> @AnoopKamath please add unit test to have code coverage pass

Done.

**Review by prgeor — COMMENTED** [review]
@AnoopKamath if you could update the cmd ref. doc https://github.com/sonic-net/sonic-utilities/blob/master/doc/Command-Reference.md

File: sfputil/main.py [hunk 4]
```diff
@@ -1526,6 +1526,43 @@ def version():
...
+        sys.exit(EXIT_FAIL)
+
+    if not is_sfp_present(port_name):
+       click.echo("{}: SFP EEPROM not detected\n".format(port_name))
+       sys.exit(EXIT_FAIL)
+
+    try:
+        api = sfp.get_xcvr_api()
+    except NotImplementedError:
+        click.echo("{}: This functionality is currently not implemented for this module".format(port_name))
```
[the remark this answers is not in the document]
Comment (AnoopKamath) [reply sfputil/main.py:1552]: Thanks @mihirpat1. taken care

**Comment by AnoopKamath** [comment]
> @AnoopKamath if you could update the cmd ref. doc https://github.com/sonic-net/sonic-utilities/blob/master/doc/Command-Reference.md

@prgeor : Taken care. Added new section 'CMIS firmware upgrade'

File: doc/Command-Reference.md [hunk 5]
```diff
@@ -2786,6 +2791,134 @@ Optionally, you can clear with a remote device name by specifying the `-d` or `-
...
+  Factory Image Version: 0.0.0
+  Running Image: B
+  Committed Image: B
+  Active Firmware: 0.3.6
+  Inactive Firmware: 0.3.5
+   ```
+
+### CMIS firmware target mode commands
+
+This command is vendor-specific and supported on the modules to set the target mode to perform remote firmware upgrades. The target modes can be set as 0 (local- E0), 1 (remote end E1), or 2 (remote end E2). Depending on the mode set, the remote or local end will respoond to CDB/I2C commands from host's E0 end. After setting the target mode, which we can use sfputil firmware upgrade commands, which will be executed on the module for whihc target mode is set.
```
Comment (prgeor) [inline doc/Command-Reference.md:2901]: @AnoopKamath for better clarity of the user can you attach a diagram showing E0, E1, E2?

Comment (AnoopKamath) [reply doc/Command-Reference.md]: taken care

[the remark this answers is not in the document]
Comment (AnoopKamath) [reply doc/Command-Reference.md]: Taken care

**Review by prgeor — APPROVED** [review]

**Comment by prgeor** [comment]
@mihirpat1 if you could help track this with ADO# for cherry pick to 202305

**Comment by mihirpat1** [comment]
> @mihirpat1 if you could help track this with ADO# for cherry pick to 202305

@StormLiangMS - Can you please help in cherry picking this to 202305?
ADO - 25473628

**Pull request merged**